### PR TITLE
Separate content/device internal states of GPUBuffer

### PR DESCRIPTION
**Pull request by kainino0x**
This change tries not to have behavior effects compared to what we have specced now, but since what we have now doesn't quite make sense, it picks just one of several possible interpretations of what we had written. #3271 is open about which one we actually pick. **I'm calling this editorial despite picking an interpretation it leaves an open issue to figure out which one to use.**

This is now independent of exposing the buffer state as an attribute (#3014). This change defines a number of content-side states from which that attribute would be defined.

The notable changes are:

- Server side now has just 2 states: available, unavailable.
- Client side has two state variables:
    - `[[mapping]]` which is either null, or all of the stuff for an actively-mapped buffer (not pending).
    - `[[map_requests]]` which tracks all map requests that haven't yet gotten a response from the device.
- At least in the mapping methods, client states are only accessed on the client, and server states are only accessed on the server.
- `unmap()` is rewritten to deinterlace the client and server side steps.
    - **`unmap()` doesn't reject promises. Instead it marks requests as cancelled so they get rejected when the map request completes.**

### CODE REVIEW

**Comment by kainino0x** [comment]
@toji PTAL, in particular that you agree this isn't making any semantic changes. I wrapped one entire function in a div so hiding whitespace might be helpful: https://github.com/gpuweb/gpuweb/pull/3300/files?w=1

@jimblandy FYI, this encodes the thing you already said was the case in #3271 (that there are multiple promises held in the state of GPUBuffer). We'll change it if that issue resolves a different way.

File: spec/index.bs [hunk 1]
```diff
@@ -3136,50 +3174,56 @@ The {{GPUMapMode}} flags determine how a {{GPUBuffer}} is mapped when calling
...
 
-                    Note: It is valid to unmap an [=invalid=] {{GPUBuffer}} that is
-                        [=buffer state/mapped at creation=] because the [=Content timeline=]
-                        might not know it is an error {{GPUBuffer}}. This allows the temporary
-                        {{GPUBuffer/[[mapping]]}} memory to be freed.
-                </div>
+                1. For each promise |p| in |this|.{{GPUBuffer/[[map_promises]]}}:
 
-            1. If |this|.{{GPUBuffer/[[state]]}} is [=buffer state/mapping pending=]:
+                    1. Reject |p| with an {{AbortError}}.
```
Comment (kainino0x) [inline spec/index.bs:3184]: OK, this isn't quite what Chromium currently does. I need to figure out if I can easily encode that behavior here, I guess. https://github.com/gpuweb/gpuweb/issues/3271#issuecomment-1210711012

**Review by toji — APPROVED** [review]
LGTM, and agreed that this doesn't make any semantic changes.

**Comment by kainino0x** [comment]
OK, one last commit. See commit message for what it did. @toji PTAL

**Comment by kainino0x** [comment]
https://github.com/gpuweb/gpuweb/pull/3300/commits/1225981e9e0f70a7efc193d11f8c16d8ba1bcc40?w=1

**Review by toji — APPROVED** [review]
Still LGTM

**Pull request merged**